File: 80b86226-8ba3-4948-affc-a059e2cc1a0c/common-lib/src/main/java/com/carbon/common/config/CacheConfig.java
package com.carbon.common.config;

import com.fasterxml.jackson.annotation.JsonTypeInfo;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.databind.jsontype.impl.LaissezFaireSubTypeValidator;
import com.fasterxml.jackson.datatype.jsr310.JavaTimeModule;
import com.github.benmanes.caffeine.cache.Caffeine;
import org.springframework.boot.autoconfigure.cache.RedisCacheManagerBuilderCustomizer;
import org.springframework.cache.CacheManager;
import org.springframework.cache.annotation.EnableCaching;
import org.springframework.cache.caffeine.CaffeineCacheManager;
import org.springframework.context.annotation.Bean;
import org.springframework.context.annotation.Configuration;
import org.springframework.context.annotation.Primary;
import org.springframework.data.redis.cache.RedisCacheConfiguration;
import org.springframework.data.redis.serializer.GenericJackson2JsonRedisSerializer;
import org.springframework.data.redis.serializer.RedisSerializationContext;
import org.springframework.data.redis.serializer.StringRedisSerializer;

import java.time.Duration;
import java.util.concurrent.TimeUnit;

@Configuration
@EnableCaching
public class CacheConfig {

    @Bean
    public CacheManager caffeineCacheManager() {
        CaffeineCacheManager cm = new CaffeineCacheManager();
        cm.setCaffeine(Caffeine.newBuilder()
                .maximumSize(50_000)
                .expireAfterWrite(30, TimeUnit.MINUTES)
                .expireAfterAccess(10, TimeUnit.MINUTES)
                .recordStats());
        cm.setAllowNullValues(false);
        return cm;
    }

    @Bean
    @Primary
    public CacheManager redisCacheManager(org.springframework.data.redis.connection.RedisConnectionFactory factory) {
        ObjectMapper om = new ObjectMapper();
        om.registerModule(new JavaTimeModule());
        om.activateDefaultTyping(LaissezFaireSubTypeValidator.instance,
                ObjectMapper.DefaultTyping.NON_FINAL, JsonTypeInfo.As.PROPERTY);
        GenericJackson2JsonRedisSerializer serializer = new GenericJackson2JsonRedisSerializer(om);

        RedisCacheConfiguration defaultCfg = RedisCacheConfiguration.defaultCacheConfig()
                .entryTtl(Duration.ofHours(2))
                .serializeKeysWith(RedisSerializationContext.SerializationPair
                        .fromSerializer(new StringRedisSerializer()))
                .serializeValuesWith(RedisSerializationContext.SerializationPair
                        .fromSerializer(serializer))
                .disableCachingNullValues();

        return org.springframework.data.redis.cache.RedisCacheManager.builder(factory)
                .cacheDefaults(defaultCfg)
                .withInitialCacheConfigurations(java.util.Map.of(
                        "factors", defaultCfg.entryTtl(Duration.ofHours(12)),
                        "emissionSources", defaultCfg.entryTtl(Duration.ofHours(6)),
                        "quotas", defaultCfg.entryTtl(Duration.ofMinutes(30))
                ))
                .transactionAware()
                .build();
    }

    @Bean
    public RedisCacheManagerBuilderCustomizer redisCacheManagerBuilderCustomizer() {
        return builder -> {
        };
    }
}

File: 9096d8e2-d12b-468a-91ae-53c97e7bf587/backend/src/main/kotlin/com/ems/dispatch/config/CacheConfig.kt
package com.ems.dispatch.config

import com.github.benmanes.caffeine.cache.Caffeine
import org.springframework.cache.CacheManager
import org.springframework.cache.annotation.EnableCaching
import org.springframework.cache.caffeine.CaffeineCacheManager
import org.springframework.context.annotation.Bean
import org.springframework.context.annotation.Configuration
import java.util.concurrent.TimeUnit

@Configuration
@EnableCaching
class CacheConfig {

    @Bean
    fun cacheManager(): CacheManager {
        val cacheManager = CaffeineCacheManager()

        cacheManager.setCacheNames(listOf(
            "ambulanceLocations",
            "ambulanceList",
            "dispatchDashboard",
            "activeEvents",
            "medicalRecords",
            "qualityMetrics",
            "vehicleStats",
            "dispatchEvent",
            "nearbyAmbulances",
            "medicalRecordDetail"
        ))

        cacheManager.setCaffeine(
            Caffeine.newBuilder()
                .maximumSize(1000)
                .expireAfterWrite(5, TimeUnit.MINUTES)
                .expireAfterAccess(2, TimeUnit.MINUTES)
                .recordStats()
        )

        return cacheManager
    }

    @Bean
    fun dashboardCacheManager(): CacheManager {
        val cacheManager = CaffeineCacheManager("dispatchDashboard")
        cacheManager.setCaffeine(
            Caffeine.newBuilder()
                .maximumSize(1)
                .expireAfterWrite(3, TimeUnit.SECONDS)
                .recordStats()
        )
        return cacheManager
    }

    @Bean
    fun ambulanceLocationCacheManager(): CacheManager {
        val cacheManager = CaffeineCacheManager("ambulanceLocations")
        cacheManager.setCaffeine(
            Caffeine.newBuilder()
                .maximumSize(200)
                .expireAfterWrite(3, TimeUnit.SECONDS)
                .recordStats()
        )
        return cacheManager
    }

    @Bean
    fun nearbyAmbulancesCacheManager(): CacheManager {
        val cacheManager = CaffeineCacheManager("nearbyAmbulances")
        cacheManager.setCaffeine(
            Caffeine.newBuilder()
                .maximumSize(100)
                .expireAfterWrite(2, TimeUnit.SECONDS)
                .recordStats()
        )
        return cacheManager
    }

    @Bean
    fun qualityMetricsCacheManager(): CacheManager {
        val cacheManager = CaffeineCacheManager("qualityMetrics")
        cacheManager.setCaffeine(
            Caffeine.newBuilder()
                .maximumSize(10)
                .expireAfterWrite(5, TimeUnit.MINUTES)
                .recordStats()
        )
        return cacheManager
    }
}
